File: frontend/apps/ios/Pantopus/Features/CeremonialMail/CeremonialMailViewModel.swift
//
//  CeremonialMailViewModel.swift
//  Pantopus
//
//  T3.7 Ceremonial Mail Compose VM — conforms to WizardModel so it
//  plugs into the shared WizardShell. Owns the four-step state
//  machine + recipient search + address verification + voice
//  postscript upload + final send.
//

import Foundation
import Observation

@Observable
@MainActor
public final class CeremonialMailViewModel: WizardModel {
    // MARK: - Public state

    public private(set) var step: CeremonialMailStep = .decide
    public private(set) var pendingEvent: CeremonialMailEvent?

    // Step 1: decide
    public var recipientQuery: String = ""
    public private(set) var recipientResults: [MailRecipientDTO] = []
    public private(set) var isSearchingRecipients: Bool = false
    public private(set) var selectedRecipient: MailRecipientDTO?
    public var intent: CeremonialMailIntent = .sayHello

    // Step 2: verify
    public private(set) var homeContext: MailHomeContextResponse?
    public var addressConfirmed: Bool = false
    public var returnAddressShared: Bool = false

    // Step 3: compose
    public var stationery: CeremonialMailStationery = .classicCream
    public var ink: CeremonialMailInk = .walnut
    public var seal: CeremonialMailSeal = .waxRed
    public var bodyText: String = ""
    public private(set) var voiceStatus: VoicePostscriptStatus = .empty

    /// Step 4: commit
    public var sendTiming: CeremonialMailSendTiming = .now

    /// Submission
    public private(set) var submitError: String?

    private let api: APIClient
    private let multipart: MultipartUploader
    private var lastSearchTask: Task<Void, Never>?
    private var isSubmitting: Bool = false

    init(
        api: APIClient = .shared,
        multipart: MultipartUploader = .shared
    ) {
        self.api = api
        self.multipart = multipart
    }

    // MARK: - Step 1: decide

    public func updateRecipientQuery(_ value: String) {
        recipientQuery = value
        if selectedRecipient != nil, value != displayName(selectedRecipient) {
            // The user is editing the field after picking — clear the
            // selection so they can pick a different recipient.
            selectedRecipient = nil
            homeContext = nil
        }
        let trimmed = value.trimmingCharacters(in: .whitespaces)
        guard trimmed.count >= 2 else {
            lastSearchTask?.cancel()
            recipientResults = []
            isSearchingRecipients = false
            return
        }
        lastSearchTask?.cancel()
        lastSearchTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: 250_000_000)
            guard !Task.isCancelled else { return }
            await self?.searchRecipients(query: trimmed)
        }
    }

    public func selectRecipient(_ recipient: MailRecipientDTO) {
        selectedRecipient = recipient
        recipientQuery = displayName(recipient)
        homeContext = nil
        addressConfirmed = false
    }

    public func selectIntent(_ value: CeremonialMailIntent) {
        intent = value
    }

    // MARK: - Step 2: verify

    public func loadHomeContext() async {
        guard let homeId = selectedRecipient?.homeId else { return }
        do {
            let response: MailHomeContextResponse =
                try await api.request(MailComposeEndpoints.homeContext(homeId: homeId))
            homeContext = response
        } catch {
            // Surfaced inline on the verify step; the user can still
            // proceed (the recipient already has a known address).
            homeContext = nil
        }
    }

    public func toggleAddressConfirmed(_ value: Bool) {
        addressConfirmed = value
    }

    public func toggleReturnAddressShared(_ value: Bool) {
        returnAddressShared = value
    }

    // MARK: - Step 3: compose

    public func selectStationery(_ value: CeremonialMailStationery) {
        stationery = value
    }

    public func selectInk(_ value: CeremonialMailInk) {
        ink = value
    }

    public func selectSeal(_ value: CeremonialMailSeal) {
        seal = value
    }

    public func updateBody(_ value: String) {
        bodyText = value
    }

    /// View-layer calls this once it has produced a recorded audio
    /// file URI. The VM kicks off the multipart upload.
    public func voicePostscriptDidRecord(localUri: String) async {
        voiceStatus = .uploading
        do {
            let url = URL(fileURLWithPath: localUri.replacingOccurrences(of: "file://", with: ""))
            let data = try Data(contentsOf: url)
            let file = MultipartFile(
                fieldName: "file",
                filename: url.lastPathComponent,
                mimeType: "audio/m4a",
                data: data
            )
            let response = try await multipart.uploadFile(
                file,
                formFields: ["file_type": "voice_postscript"]
            )
            voiceStatus = .uploaded(remoteUrl: response.file.url)
        } catch {
            voiceStatus = .error(message: "Couldn't upload postscript. Try again.")
        }
    }

    public func clearVoicePostscript() {
        voiceStatus = .empty
    }

    public func voicePostscriptDidStartRecording() {
        voiceStatus = .recording
    }

    /// Mark the recording as captured before the upload kicks off.
    /// The view sets the URI; the upload then runs through
    /// `voicePostscriptDidRecord`.
    public func voicePostscriptDidCapture(localUri: String) {
        voiceStatus = .captured(localUri: localUri)
    }

    // MARK: - Step 4: commit

    public func selectSendTiming(_ value: CeremonialMailSendTiming) {
        sendTiming = value
    }

    // MARK: - WizardModel

    public var chrome: WizardChrome {
        WizardChrome(
            title: stepTitle,
            progressLabel: step.stepNumber.map { .stepOf(current: $0, total: CeremonialMailStep.progressTotal) }
                ?? .hidden,
            progressFraction: step.stepNumber.map { Double($0) / Double(CeremonialMailStep.progressTotal) },
            leading: step == .decide ? .close : .back,
            primaryCTALabel: primaryCtaLabel,
            primaryCTAEnabled: primaryCtaEnabled,
            secondaryCTA: step == .success
                ? WizardSecondaryCTA(label: "Back to Hub", identifier: "ceremonialBackToHub")
                : nil,
            isSubmitting: isSubmitting,
            dirty: stepDirty,
            showsProgressBar: step != .success
        )
    }

    public func leadingTapped() {
        guard step != .decide else { pendingEvent = .dismiss
            return
        }
        switch step {
        case .verify: step = .decide
        case .compose: step = .verify
        case .commit: step = .compose
        case .success: pendingEvent = .dismiss
        case .decide: break
        }
    }

    public func discardConfirmed() {
        pendingEvent = .dismiss
    }

    public func primaryTapped() {
        switch step {
        case .decide:
            guard selectedRecipient != nil else { return }
            step = .verify
            Task { await loadHomeContext() }
        case .verify:
            guard addressConfirmed else { return }
            step = .compose
        case .compose:
            guard !bodyText.trimmingCharacters(in: .whitespaces).isEmpty else { return }
            step = .commit
        case .commit:
            guard !isSubmitting else { return }
            Task { await submit() }
        case .success:
            pendingEvent = .dismiss
        }
    }

    public func secondaryTapped() {
        if step == .success {
            pendingEvent = .dismiss
        }
    }

    public func acknowledgePendingEvent() {
        pendingEvent = nil
    }

    // MARK: - Computed

    private var stepTitle: String {
        // T6.5e — Step titles match the ceremonial-compose design
        // (Porch Call · Address It · Write It · Seal & Send) while the
        // internal enum stays decide / verify / compose / commit so
        // existing tests and analytics events keep their stable names.
        switch step {
        case .decide: "Porch Call"
        case .verify: "Address It"
        case .compose: "Write It"
        case .commit: "Seal & Send"
        case .success: "Letter on its way"
        }
    }

    private var primaryCtaLabel: String {
        switch step {
        case .decide: "Continue"
        case .verify: "Continue to letter"
        case .compose: "Continue"
        case .commit: "Seal and send"
        case .success: "Done"
        }
    }

    private var primaryCtaEnabled: Bool {
        switch step {
        case .decide: selectedRecipient != nil
        case .verify: addressConfirmed
        case .compose: !bodyText.trimmingCharacters(in: .whitespaces).isEmpty
        case .commit: !isSubmitting
        case .success: true
        }
    }

    private var stepDirty: Bool {
        switch step {
        case .decide: selectedRecipient != nil || !recipientQuery.isEmpty
        case .verify, .compose, .commit: true
        case .success: false
        }
    }

    // MARK: - Network

    private func searchRecipients(query: String) async {
        isSearchingRecipients = true
        defer { isSearchingRecipients = false }
        do {
            let response: MailComposeRecipientsResponse =
                try await api.request(MailComposeEndpoints.recipients(query: query))
            recipientResults = response.recipients
        } catch {
            recipientResults = []
        }
    }

    private func submit() async {
        guard let recipient = selectedRecipient else { return }
        isSubmitting = true
        defer { isSubmitting = false }
        submitError = nil
        let voiceRemote: String? = {
            if case let .uploaded(remoteUrl) = voiceStatus { return remoteUrl }
            return nil
        }()
        let body = SendMailBody(
            recipientUserId: recipient.userId,
            recipientHomeId: recipient.homeId,
            type: "letter",
            subject: subjectFromIntent(),
            content: bodyText.trimmingCharacters(in: .whitespacesAndNewlines),
            object: SendMailObject(
                title: subjectFromIntent(),
                content: bodyText.trimmingCharacters(in: .whitespacesAndNewlines),
                payload: SendMailPayload(
                    stationeryTheme: stationery.rawValue,
                    inkSelection: ink.rawValue,
                    sealChoice: seal.rawValue,
                    intent: intent.rawValue,
                    returnAddressShared: returnAddressShared,
                    voicePostscriptUri: voiceRemote
                )
            ),
            expiresAt: nil
        )
        do {
            let response: SendMailResponse = try await api.request(MailComposeEndpoints.send(body: body))
            step = .success
            if let mailId = response.mail?.id {
                pendingEvent = .openMail(mailId: mailId)
            }
        } catch {
            submitError = (error as? APIError)?.errorDescription ?? "Couldn't send your letter. Try again."
        }
    }

    // MARK: - Helpers

    private func subjectFromIntent() -> String {
        switch intent {
        case .sayHello: "A note from a friend"
        case .congratulations: "Congratulations!"
        case .condolences: "Thinking of you"
        case .businessNote: "A business note"
        case .justBecause: "Just because"
        }
    }

    private func displayName(_ recipient: MailRecipientDTO?) -> String {
        guard let recipient else { return "" }
        return recipient.name ?? recipient.username ?? "Recipient"
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/ceremonial_mail/CeremonialMailViewModel.kt
@file:Suppress("PackageNaming", "LongMethod", "TooManyFunctions", "MagicNumber")

package app.pantopus.android.ui.screens.ceremonial_mail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.mail_compose.MailHomeContextResponse
import app.pantopus.android.data.api.models.mail_compose.MailRecipientDto
import app.pantopus.android.data.api.models.mail_compose.SendMailBody
import app.pantopus.android.data.api.models.mail_compose.SendMailObject
import app.pantopus.android.data.api.models.mail_compose.SendMailPayload
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.mail_compose.MailComposeRepository
import app.pantopus.android.ui.screens.shared.wizard.WizardChrome
import app.pantopus.android.ui.screens.shared.wizard.WizardLeadingControl
import app.pantopus.android.ui.screens.shared.wizard.WizardModel
import app.pantopus.android.ui.screens.shared.wizard.WizardProgressLabel
import app.pantopus.android.ui.screens.shared.wizard.WizardSecondaryCta
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class CeremonialMailViewModel
    @Inject
    constructor(
        private val repository: MailComposeRepository,
    ) : ViewModel(), WizardModel {
        private val _form = MutableStateFlow(CeremonialMailFormState())
        val form: StateFlow<CeremonialMailFormState> = _form.asStateFlow()

        private val _recipientResults = MutableStateFlow<List<MailRecipientDto>>(emptyList())
        val recipientResults: StateFlow<List<MailRecipientDto>> = _recipientResults.asStateFlow()

        private val _selectedRecipient = MutableStateFlow<MailRecipientDto?>(null)
        val selectedRecipient: StateFlow<MailRecipientDto?> = _selectedRecipient.asStateFlow()

        private val _homeContext = MutableStateFlow<MailHomeContextResponse?>(null)
        val homeContext: StateFlow<MailHomeContextResponse?> = _homeContext.asStateFlow()

        private val _isSearching = MutableStateFlow(false)
        val isSearching: StateFlow<Boolean> = _isSearching.asStateFlow()

        private val _isSubmitting = MutableStateFlow(false)
        val isSubmitting: StateFlow<Boolean> = _isSubmitting.asStateFlow()

        private val _submitError = MutableStateFlow<String?>(null)
        val submitError: StateFlow<String?> = _submitError.asStateFlow()

        private val _voiceStatus = MutableStateFlow<VoicePostscriptStatus>(VoicePostscriptStatus.Empty)
        val voiceStatus: StateFlow<VoicePostscriptStatus> = _voiceStatus.asStateFlow()

        private val _pendingEvent = MutableStateFlow<CeremonialMailEvent?>(null)
        val pendingEvent: StateFlow<CeremonialMailEvent?> = _pendingEvent.asStateFlow()

        private var searchJob: Job? = null

        // MARK: - Step 1: decide

        fun updateRecipientQuery(value: String) {
            _form.value = _form.value.copy(recipientQuery = value)
            val current = _selectedRecipient.value
            if (current != null && value != displayName(current)) {
                _selectedRecipient.value = null
                _homeContext.value = null
            }
            val trimmed = value.trim()
            searchJob?.cancel()
            if (trimmed.length < 2) {
                _recipientResults.value = emptyList()
                _isSearching.value = false
                return
            }
            searchJob =
                viewModelScope.launch {
                    delay(250)
                    searchRecipients(trimmed)
                }
        }

        fun selectRecipient(recipient: MailRecipientDto) {
            _selectedRecipient.value = recipient
            _form.value = _form.value.copy(recipientQuery = displayName(recipient), addressConfirmed = false)
            _homeContext.value = null
        }

        fun selectIntent(intent: CeremonialMailIntent) {
            _form.value = _form.value.copy(intent = intent)
        }

        // MARK: - Step 2: verify

        fun loadHomeContext() {
            val homeId = _selectedRecipient.value?.homeId ?: return
            viewModelScope.launch {
                when (val result = repository.homeContext(homeId)) {
                    is NetworkResult.Success -> _homeContext.value = result.data
                    is NetworkResult.Failure -> _homeContext.value = null
                }
            }
        }

        fun toggleAddressConfirmed(value: Boolean) {
            _form.value = _form.value.copy(addressConfirmed = value)
        }

        fun toggleReturnAddressShared(value: Boolean) {
            _form.value = _form.value.copy(returnAddressShared = value)
        }

        // MARK: - Step 3: compose

        fun selectStationery(value: CeremonialMailStationery) {
            _form.value = _form.value.copy(stationery = value)
        }

        fun selectInk(value: CeremonialMailInk) {
            _form.value = _form.value.copy(ink = value)
        }

        fun selectSeal(value: CeremonialMailSeal) {
            _form.value = _form.value.copy(seal = value)
        }

        fun updateBody(value: String) {
            _form.value = _form.value.copy(bodyText = value)
        }

        fun voicePostscriptDidStartRecording() {
            _voiceStatus.value = VoicePostscriptStatus.Recording
        }

        /** View calls this after capturing audio. We accept the
         *  local URI; the upload helper that wraps `/api/files/upload`
         *  is host-provided so tests can inject a mock. */
        fun voicePostscriptDidCapture(localUri: String) {
            _voiceStatus.value = VoicePostscriptStatus.Captured(localUri)
        }

        /** Apply an upload result from the platform helper. */
        fun voicePostscriptDidUpload(remoteUrl: String) {
            _voiceStatus.value = VoicePostscriptStatus.Uploaded(remoteUrl)
        }

        fun voicePostscriptDidFail(message: String) {
            _voiceStatus.value = VoicePostscriptStatus.Error(message)
        }

        fun clearVoicePostscript() {
            _voiceStatus.value = VoicePostscriptStatus.Empty
        }

        // MARK: - Step 4: commit

        fun selectSendTiming(value: CeremonialMailSendTiming) {
            _form.value = _form.value.copy(sendTiming = value)
        }

        fun submit() {
            if (_isSubmitting.value) return
            val recipient = _selectedRecipient.value ?: return
            val form = _form.value
            _isSubmitting.value = true
            _submitError.value = null
            val voiceUri =
                (_voiceStatus.value as? VoicePostscriptStatus.Uploaded)?.remoteUrl
            val body =
                SendMailBody(
                    recipientUserId = recipient.userId,
                    recipientHomeId = recipient.homeId,
                    type = "letter",
                    subject = subjectFromIntent(form.intent),
                    content = form.bodyText.trim(),
                    `object` =
                        SendMailObject(
                            title = subjectFromIntent(form.intent),
                            content = form.bodyText.trim(),
                            payload =
                                SendMailPayload(
                                    stationeryTheme = form.stationery.wire,
                                    inkSelection = form.ink.wire,
                                    sealChoice = form.seal.wire,
                                    intent = form.intent.wire,
                                    returnAddressShared = form.returnAddressShared,
                                    voicePostscriptUri = voiceUri,
                                ),
                        ),
                    expiresAt = null,
                )
            viewModelScope.launch {
                when (val result = repository.send(body)) {
                    is NetworkResult.Success -> {
                        _isSubmitting.value = false
                        _form.value = form.copy(step = CeremonialMailStep.Success)
                        result.data.mail?.id?.let { _pendingEvent.value = CeremonialMailEvent.OpenMail(it) }
                    }
                    is NetworkResult.Failure -> {
                        _isSubmitting.value = false
                        _submitError.value = "Couldn't send your letter. Try again."
                    }
                }
            }
        }

        fun acknowledgePendingEvent() {
            _pendingEvent.value = null
        }

        // MARK: - WizardModel

        override val chrome: WizardChrome
            get() {
                val current = _form.value
                val total = CeremonialMailStep.PROGRESS_TOTAL
                val number = current.step.stepNumber
                return WizardChrome(
                    title = titleFor(current.step),
                    progressLabel =
                        number?.let { WizardProgressLabel.StepOf(it, total) }
                            ?: WizardProgressLabel.Hidden,
                    progressFraction = number?.toFloat()?.div(total.toFloat()),
                    leading = if (current.step == CeremonialMailStep.Decide) WizardLeadingControl.Close else WizardLeadingControl.Back,
                    primaryCtaLabel = primaryCtaLabelFor(current.step),
                    primaryCtaEnabled = primaryEnabledFor(current),
                    secondaryCta =
                        if (current.step == CeremonialMailStep.Success) {
                            WizardSecondaryCta(label = "Back to Hub", testTag = "ceremonialBackToHub")
                        } else {
                            null
                        },
                    isSubmitting = _isSubmitting.value,
                    dirty = dirtyFor(current),
                    showsProgressBar = current.step != CeremonialMailStep.Success,
                )
            }

        override fun onLeading() {
            val step = _form.value.step
            if (step == CeremonialMailStep.Decide) {
                _pendingEvent.value = CeremonialMailEvent.Dismiss
                return
            }
            _form.value =
                _form.value.copy(
                    step =
                        when (step) {
                            CeremonialMailStep.Verify -> CeremonialMailStep.Decide
                            CeremonialMailStep.Compose -> CeremonialMailStep.Verify
                            CeremonialMailStep.Commit -> CeremonialMailStep.Compose
                            CeremonialMailStep.Success -> CeremonialMailStep.Commit
                            CeremonialMailStep.Decide -> CeremonialMailStep.Decide
                        },
                )
            if (step == CeremonialMailStep.Success) _pendingEvent.value = CeremonialMailEvent.Dismiss
        }

        override fun onDiscard() {
            _pendingEvent.value = CeremonialMailEvent.Dismiss
        }

        override fun onPrimary() {
            val step = _form.value.step
            when (step) {
                CeremonialMailStep.Decide -> {
                    if (_selectedRecipient.value == null) return
                    _form.value = _form.value.copy(step = CeremonialMailStep.Verify)
                    loadHomeContext()
                }
                CeremonialMailStep.Verify -> {
                    if (!_form.value.addressConfirmed) return
                    _form.value = _form.value.copy(step = CeremonialMailStep.Compose)
                }
                CeremonialMailStep.Compose -> {
                    if (_form.value.bodyText.trim().isEmpty()) return
                    _form.value = _form.value.copy(step = CeremonialMailStep.Commit)
                }
                CeremonialMailStep.Commit -> submit()
                CeremonialMailStep.Success -> _pendingEvent.value = CeremonialMailEvent.Dismiss
            }
        }

        override fun onSecondary() {
            if (_form.value.step == CeremonialMailStep.Success) {
                _pendingEvent.value = CeremonialMailEvent.Dismiss
            }
        }

        // MARK: - Helpers

        private fun titleFor(step: CeremonialMailStep): String =
            // T6.5e — Step titles match the ceremonial-compose design
            // (Porch Call · Address It · Write It · Seal & Send) while
            // the internal enum stays decide / verify / compose / commit
            // so existing tests and analytics events keep their stable
            // names.
            when (step) {
                CeremonialMailStep.Decide -> "Porch Call"
                CeremonialMailStep.Verify -> "Address It"
                CeremonialMailStep.Compose -> "Write It"
                CeremonialMailStep.Commit -> "Seal & Send"
                CeremonialMailStep.Success -> "Letter on its way"
            }

        private fun primaryCtaLabelFor(step: CeremonialMailStep): String =
            when (step) {
                CeremonialMailStep.Decide -> "Continue"
                CeremonialMailStep.Verify -> "Continue to letter"
                CeremonialMailStep.Compose -> "Continue"
                CeremonialMailStep.Commit -> "Seal and send"
                CeremonialMailStep.Success -> "Done"
            }

        private fun primaryEnabledFor(form: CeremonialMailFormState): Boolean =
            when (form.step) {
                CeremonialMailStep.Decide -> _selectedRecipient.value != null
                CeremonialMailStep.Verify -> form.addressConfirmed
                CeremonialMailStep.Compose -> form.bodyText.trim().isNotEmpty()
                CeremonialMailStep.Commit -> !_isSubmitting.value
                CeremonialMailStep.Success -> true
            }

        private fun dirtyFor(form: CeremonialMailFormState): Boolean =
            when (form.step) {
                CeremonialMailStep.Decide -> _selectedRecipient.value != null || form.recipientQuery.isNotEmpty()
                CeremonialMailStep.Verify, CeremonialMailStep.Compose, CeremonialMailStep.Commit -> true
                CeremonialMailStep.Success -> false
            }

        private fun subjectFromIntent(intent: CeremonialMailIntent): String =
            when (intent) {
                CeremonialMailIntent.SayHello -> "A note from a friend"
                CeremonialMailIntent.Congratulations -> "Congratulations!"
                CeremonialMailIntent.Condolences -> "Thinking of you"
                CeremonialMailIntent.BusinessNote -> "A business note"
                CeremonialMailIntent.JustBecause -> "Just because"
            }

        private fun displayName(recipient: MailRecipientDto?): String = recipient?.name ?: recipient?.username ?: ""

        private suspend fun searchRecipients(query: String) {
            _isSearching.value = true
            try {
                when (val result = repository.recipients(query)) {
                    is NetworkResult.Success -> _recipientResults.value = result.data.recipients
                    is NetworkResult.Failure -> _recipientResults.value = emptyList()
                }
            } finally {
                _isSearching.value = false
            }
        }
    }
